packages/devtools_app/lib/src/shared/primitives/utils.dart

```diff
@@ -442,84 +442,101 @@ class RateLimiter {
-  }) {
-    final originalStart = original.start;
-    final originalEnd = original.end;
-    return TimeRange()
-      ..start = originalStart != null ? originalStart + offset : null
-      ..end = originalEnd != null ? originalEnd + offset : null;
+  int? _start;
+
+  /// Sets the start time of this builder.
+  ///
+  /// The start time should be less than or equal to the end time.
+  set start(int startTime) {
+    _start = startTime;
   }
 
-  final bool singleAssignment;
+  int? _end;
 
-  Duration? get start => _start;
+  /// Sets the end time of this builder.
+  ///
+  /// The end time should be greater than or equal to the start time.
+  set end(int endTime) {
+    _end = endTime;
+  }
 
-  Duration? _start;
+  /// Whether both `start` and `end` properties are set,
+  /// meaning [build] can safely be called.
+  bool get canBuild => _start != null && _end != null;
 
-  set start(Duration? value) {
-    if (singleAssignment) {
-      assert(_start == null);
+  /// Returns a [TimeRange] built from the specified [start] and [end] values.
+  ///
+  /// If either [start] or [end] is `null`, throws an error.
+  ///
+  /// The [start] time must be less than or equal to the [end] time.
+  TimeRange build() {
+    final startDuration = _start;
+    if (startDuration == null) {
+      throw StateError('TimeRangeBuilder.start must be set before building!');
     }
-    if (value != null && _end != null) {
-      assert(
-        value <= _end!,
-        '$value is not less than or equal to end time $_end',
-      );
+
+    final endDuration = _end;
+    if (endDuration == null) {
+      throw StateError('TimeRangeBuilder.end must be set before building!');
     }
-    _start = value;
-  }
 
-  Duration? get end => _end;
+    return TimeRange(start: startDuration, end: endDuration);
+  }
 
-  Duration? _end;
+  /// Returns a new [TimeRangeBuilder] with the current values of this builder.
+  TimeRangeBuilder copy() => TimeRangeBuilder(start: _start, end: _end);
+}
 
-  set end(Duration? value) {
-    if (singleAssignment) {
-      assert(_end == null);
-    }
-    if (value != null && _start != null) {
+final class TimeRange {
+  /// Creates a [TimeRange] with the specified
+  /// [start] and [end] times in microseconds.
+  ///
+  /// The [start] time must be less than or equal to the [end] time.
+  TimeRange({required this.start, required this.end})
+    : assert(start <= end, '$start is not less than or equal to end time $end'),
       assert(
-        value >= _start!,
-        '$value is not greater than or equal to start time $_start',
+        end >= start,
+        '$end is not greater than or equal to start time $start',
       );
-    }
-    _end = value;
-  }
 
-  Duration get duration => end! - start!;
+  /// Creates a [TimeRange] with the specified [start] time in microseconds and
+  /// [end] calculated as being [length] microseconds later.
+  factory TimeRange.ofLength({int start = 0, required int length}) =>
```

Comment:
nit: can we have the required param before the optional one in the list?

I also wonder if `TimeRange.ofDuration` is a clearer name

Reply:
Good point! `ofDuration` does read better and makes `length`/`duration` work well as a required positional parameter instead of named.